File: java/tests/StrataPersistenceTest/src/strata/persistence/testdomain/PersonRepositoryTest.java
package strata.persistence.testdomain;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import java.util.Currency;
import java.util.List;
import strata.foundation.value.DateTime;
import strata.foundation.value.Money;
import strata.persistence.namedquery.INamedQuery;
import strata.persistence.unitofwork.IUnitOfWork;
import strata.persistence.unitofwork.IUnitOfWorkProvider;

public abstract
class PersonRepositoryTest
{
    private IUnitOfWorkProvider itsProvider;
    private IPersonRepository   itsTarget;
    
    @Before
    public void 
    setUp() throws Exception
    {
        itsProvider = createUnitOfWorkProvider();
        itsTarget = new PersonRepository(itsProvider);
    }

    @After
    public void 
    tearDown() throws Exception
    {
        INamedQuery<IPerson> finder     = itsTarget.getNamedQuery( "GetAll" );
        IUnitOfWork          unitOfWork = itsProvider.getUnitOfWork();
        
        for (IPerson p : finder.getAll())
            itsTarget.removePerson( p );
        
        unitOfWork.commit();
        itsTarget = null;
        itsProvider = null;
    }

    @Test
    public void 
    testInsertPerson() 
        throws Exception
    {
        IUnitOfWork unitOfWork = itsProvider.getUnitOfWork();
        IPerson     expected = 
                        new Person(
                            new PersonName("John","Liebenau"),
                            new PersonAge(
                                new DateTime(1967,4,25)));
        IPerson     actual = 
                        itsTarget.insertPerson( expected );
        
        unitOfWork.commit();
        Assert.assertEquals( expected.getName(),actual.getName() );
        Assert.assertEquals( expected.getAge(),actual.getAge() );
    }

    @Test
    public void 
    testUpdatePerson() 
        throws Exception
    {
        IUnitOfWork unitOfWork = itsProvider.getUnitOfWork();
        IPerson     expected = 
                        new Person(
                            new PersonName("John","Liebenau"),
                            new PersonAge(
                                new DateTime(1967,4,25)));
        IPerson     actual = null;
        
        
        expected = itsTarget.insertPerson( expected );
        unitOfWork.commit();
        
        unitOfWork = itsProvider.getUnitOfWork();
        expected.getName().setMiddleName( "Friedrich" );
        expected.setNetWorth( new Money( Currency.getInstance( "EUR" ),100.00) );
        actual = itsTarget.updatePerson( expected );
        unitOfWork.commit();
        Assert.assertEquals( expected.getName(),actual.getName() );
        Assert.assertEquals( expected.getAge(),actual.getAge() );
    }

    @Test
    public void 
    testRemovePerson() 
        throws Exception
    {
        IUnitOfWork unitOfWork = itsProvider.getUnitOfWork();
        IPerson     expected = 
                        new Person(
                            new PersonName("John","Liebenau"),
                            new PersonAge(new DateTime(1967,4,25)));

        expected = itsTarget.insertPerson( expected );
        unitOfWork.commit();
        
        unitOfWork = itsProvider.getUnitOfWork();
        Assert.assertTrue( itsTarget.hasPersonWithPartyKey( expected.getPartyKey() ) );
        itsTarget.removePerson( expected );
        unitOfWork.commit();
        
        Assert.assertFalse( itsTarget.hasPersonWithPartyKey( expected.getPartyKey() ) );
    }

    @Test
    public void 
    testGetPersonByPartyKey()
        throws Exception
    {
        IUnitOfWork unitOfWork = itsProvider.getUnitOfWork();
        IPerson     expected = 
                        new Person(
                            new PersonName("John","Liebenau"),
                            new PersonAge(new DateTime(1967,4,25)));
        IPerson     actual = null;
        
        expected = itsTarget.insertPerson( expected );
        unitOfWork.commit();
        
        Assert.assertTrue( itsTarget.hasPersonWithPartyKey( expected.getPartyKey() ) );
        actual = itsTarget.getPersonByPartyKey( expected.getPartyKey() );
        Assert.assertEquals( expected,actual );
    }

    @Test
    public void 
    testGetPersonByName() 
        throws Exception
    {
        IUnitOfWork unitOfWork = itsProvider.getUnitOfWork();
        IPerson       expected = 
                          new Person(
                              new PersonName("John","Liebenau"),
                              new PersonAge(new DateTime(1967,4,25)));
        List<IPerson> actual = null;
        
        expected = itsTarget.insertPerson( expected );
        unitOfWork.commit();
        
        Assert.assertTrue( itsTarget.hasPersonWithPartyKey( expected.getPartyKey() ) );
        actual = itsTarget.getPersonsByName( expected.getName() );
        Assert.assertEquals( 1,actual.size() );
        Assert.assertEquals( expected,actual.get( 0 ) );
        
    }

    @Test
    public void 
    testGetFinder()
    {
        INamedQuery<IPerson> query = itsTarget.getNamedQuery( "GetByName" );
        
        Assert.assertNotNull( query );
        Assert.assertEquals( "strata.persistence.testdomain.IPerson.GetByName",query.getName() );
    }

    @Test
    public void 
    testHasPersonWithPartyKey() 
        throws Exception
    {
        IUnitOfWork unitOfWork = itsProvider.getUnitOfWork();
        IPerson     expected = 
                        new Person(
                            new PersonName("John","Liebenau"),
                            new PersonAge(new DateTime(1967,4,25)));
        IPerson     actual = null;
        
        expected = itsTarget.insertPerson( expected );
        unitOfWork.commit();
        
        Assert.assertTrue( itsTarget.hasPersonWithPartyKey( expected.getPartyKey() ) );
    }

    @Ignore
    @Test
    public void 
    testHasPersonWithName() 
        throws Exception
    {
        IUnitOfWork unitOfWork = itsProvider.getUnitOfWork();
        IPerson     expected = 
                          new Person(
                              new PersonName("John","Liebenau"),
                              new PersonAge(new DateTime(1967,4,25)));
        
        expected = itsTarget.insertPerson( expected );
        unitOfWork.commit();
        
        Assert.assertTrue( itsTarget.hasPersonWithName( expected.getName() ) );
        unitOfWork = itsProvider.getUnitOfWork();
        itsTarget.removePerson( expected );
        unitOfWork.commit();
        Assert.assertFalse( itsTarget.hasPersonWithName( expected.getName() ) );
    }

    @Test
    public void 
    testHasFinder()
    {
        Assert.assertTrue( itsTarget.hasNamedQuery( "GetByName" ) );
    }

    protected abstract IUnitOfWorkProvider
    createUnitOfWorkProvider();
}

// ##########################################################################

File: csharp/Tests/Strata.Domain.CoreTest/Repository/PersonRepositoryTest.cs
//  ##########################################################################
//  # File Name: PersonRepositoryTest.cs
//  ##########################################################################

using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using NUnit.Framework;
using Strata.Domain.Core.NamedQuery;
using Strata.Domain.Core.TestDomain;
using Strata.Domain.Core.UnitOfWork;
using Strata.Foundation.Core.Value;

namespace Strata.Domain.Core.Repository
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// </summary>
    ///
    [TestFixture]
    public abstract
    class PersonRepositoryTest
    {
        private IUnitOfWorkProvider itsProvider;
        private IPersonRepository itsTarget;

        [SetUp]
        public void
        SetUp()
        {
            itsProvider = CreateUnitOfWorkProvider();
            itsTarget = new PersonRepository(itsProvider);
        }

        [TearDown]
        public async Task
        TearDown()
        {
            INamedQuery<IPerson> finder = await (itsTarget.GetNamedQuery("GetAll"));
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();

            foreach (IPerson p in await (finder.GetAll()))
                await (itsTarget.Remove(p));

            await (unitOfWork.Commit());
            itsTarget = null;
            itsProvider = null;
        }

        [Test]
        public async Task
        TestInsertPerson()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IPerson expected =
                            new Person()
                            {
                                Name = new PersonName("John","Liebenau"),
                                DateOfBirth = new DateTime(1967,4,25)
                            };

            IPerson actual = await (itsTarget.Insert(expected));

            await (unitOfWork.Commit());
            Assert.AreEqual(expected.Name,actual.Name);
            Assert.AreEqual(expected.DateOfBirth,actual.DateOfBirth);
        }

        [Test]
        public async Task
        TestUpdatePerson()

        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IPerson expected =
                new Person()
                {
                    Name = new PersonName("John","Liebenau"),
                    DateOfBirth = new DateTime(1967,4,25)
                };
            IPerson actual = null;


            expected = await(itsTarget.Insert( expected ));
            await(unitOfWork.Commit());

            unitOfWork = itsProvider.GetUnitOfWork();
            expected.Name.MiddleName = "Friedrich";
            actual = await(itsTarget.Update( expected ));
            await(unitOfWork.Commit());

            Assert.AreEqual(expected.Name,actual.Name);
            Assert.AreEqual(expected.DateOfBirth,actual.DateOfBirth);
            
        }

        [Test]
        public async Task
        TestRemovePerson()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IPerson expected =
                new Person()
                {
                    Name = new PersonName("John","Liebenau"),
                    DateOfBirth = new DateTime(1967,4,25)
                };
            IPerson actual = null;


            expected = await (itsTarget.Insert(expected));
            await (unitOfWork.Commit());

            unitOfWork = itsProvider.GetUnitOfWork();

            Assert.IsTrue(await (itsTarget.HasUnique(expected.PrimaryId)));
            await (itsTarget.Remove(expected));
            await (unitOfWork.Commit());

            Assert.IsFalse(await (itsTarget.HasUnique(expected.PrimaryId)));
        }

        [Test]
        public async Task
        TestGetPersonByPartyKey()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IPerson expected =
                new Person()
                {
                    Name = new PersonName("John","Liebenau"),
                    DateOfBirth = new DateTime(1967,4,25)
                };
            IPerson actual = null;


            expected = await(itsTarget.Insert(expected));
            await(unitOfWork.Commit());

            unitOfWork = itsProvider.GetUnitOfWork();
            Assert.IsTrue(await (itsTarget.HasUnique(expected.PrimaryId)));
            actual = (await itsTarget.GetUnique(expected.PrimaryId)).Get();
            Assert.AreEqual(expected,actual);
        }

        [Test]
        public async Task
        TestGetPersonByName()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IPerson expected =
                new Person()
                {
                    Name = new PersonName("John","Liebenau"),
                    DateOfBirth = new DateTime(1967,4,25)
                };
            IPerson actual = null;


            expected = await (itsTarget.Insert(expected));
            await (unitOfWork.Commit());

            unitOfWork = itsProvider.GetUnitOfWork();
            Assert.IsTrue(await (itsTarget.HasUnique(expected.PrimaryId)));
            actual = (await itsTarget.GetUniqueMatching(entity => entity.Name.Equals(expected.Name))).Get();
            Assert.AreEqual(expected,actual);
        }

        [Test]
        public async Task
        TestGetFinder()
        {
            INamedQuery<IPerson> query = await (itsTarget.GetNamedQuery("GetByName"));

            Assert.NotNull(query);
            Assert.AreEqual("GetByName",query.Name);
        }

        [Test]
        public async Task
        TestHasPersonWithPartyKey()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IPerson expected =
                new Person()
                {
                    Name = new PersonName("John","Liebenau"),
                    DateOfBirth = new DateTime(1967,4,25)
                };
            IPerson actual = null;


            expected = await (itsTarget.Insert(expected));
            await (unitOfWork.Commit());

            unitOfWork = itsProvider.GetUnitOfWork();
            Assert.IsTrue(await (itsTarget.HasUnique(expected.PrimaryId)));
        }

        [Ignore("")]
        [Test]
        public async Task
        TestHasPersonWithName()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IPerson expected =
                new Person()
                {
                    Name = new PersonName("John","Liebenau"),
                    DateOfBirth = new DateTime(1967,4,25)
                };
            IPerson actual = null;


            expected = await(itsTarget.Insert(expected));
            await(unitOfWork.Commit());

            unitOfWork = itsProvider.GetUnitOfWork();
            Assert.IsTrue(await(itsTarget.HasMatching(entity => entity.Name.Equals(expected.Name))));
        }

        [Test]
        public async Task
        TestHasFinder()
        {
            Assert.IsTrue(await (itsTarget.HasNamedQuery("GetByName")));
        }

        protected abstract IUnitOfWorkProvider
        CreateUnitOfWorkProvider();
    }
}

//  ##########################################################################
